src/core/components/pin-set.js

```diff
@@ -0,0 +1,269 @@
+const defaultFanout = 256
+const maxItems = 8192
+const pb = protobuf(pbSchema)
+
+function toB58String (hash) {
+  return new CID(hash).toBaseEncodedString()
+}
+
+function readHeader (rootNode) {
+  // rootNode.data should be a buffer of the format:
+  // < varint(headerLength) | header | itemData... >
+  const rootData = rootNode.data
+  const hdrLength = varint.decode(rootData)
+  const vBytes = varint.decode.bytes
+  if (vBytes <= 0) {
+    throw new Error('Invalid Set header length')
+  }
+  if (vBytes + hdrLength > rootData.length) {
+    throw new Error('Impossibly large set header length')
+  }
+  const hdrSlice = rootData.slice(vBytes, hdrLength + vBytes)
+  const header = pb.Set.decode(hdrSlice)
+  if (header.version !== 1) {
+    throw new Error(`Unsupported Set version: ${header.version}`)
+  }
+  if (header.fanout > rootNode.links.length) {
+    throw new Error('Impossibly large fanout')
+  }
+  return {
+    header: header,
+    data: rootData.slice(hdrLength + vBytes)
+  }
+}
+
+function hash (seed, key) {
+  const buf = Buffer.alloc(4)
+  buf.writeUInt32LE(seed, 0)
+  const data = Buffer.concat([
+    buf, Buffer.from(toB58String(key))
+  ])
+  return fnv1a(data.toString('binary'))
+}
+
+exports = module.exports = function (dag) {
+  const pinSet = {
+    // should this be part of `object` API?
+    hasChild: (root, childhash, callback) => {
+      const seen = {}
+      if (CID.isCID(childhash) || Buffer.isBuffer(childhash)) {
+        childhash = toB58String(childhash)
+      }
+
+      function searchChildren (root, cb, numToCheck, numChecked) {
+        if (!root.links.length && numToCheck === numChecked) {
+          // all nodes have been checked
+          return cb(null, false)
+        }
+
+        some(root.links, ({ multihash }, someCb) => {
+          const bs58Link = toB58String(multihash)
+          if (bs58Link in seen) return
+          if (bs58Link === childhash) {
+            return someCb(null, true)
+          }
+
+          seen[bs58Link] = true
+
+          dag.get(multihash, (err, { value }) => {
+            if (err) { return someCb(err) }
+
+            numChecked++
+            numToCheck += value.links.length
+            searchChildren(value, someCb, numToCheck, numChecked)
+          })
+        }, cb)
+      }
+
+      return searchChildren(root, callback, root.links.length, 0)
+    },
+
+    storeSet: (keys, logInternalKey, callback) => {
+      const pins = keys.map(key => ({
+        key: key,
+        data: null
+      }))
+
+      pinSet.storeItems(pins, logInternalKey, (err, rootNode) => {
+        if (err) { return callback(err) }
+        const opts = { cid: new CID(rootNode.multihash) }
+        dag.put(rootNode, opts, (err, cid) => {
+          if (err) { return callback(err) }
+          logInternalKey(rootNode.multihash)
+          callback(null, rootNode)
+        })
+      })
+    },
+
+    storeItems: (pins, logInternalKey, callback, _depth = 0, _binsToFill = 0, _binsFilled = 0) => {
```

Comment:
I think I've left a similar comment before, but can't find it anymore. So the underscore means optional arguments? I haven't seen this convention throughout the JS projects yet. I normally use underscore for variables not being used.

I just want to have that said, feel free to keep it as it currently is.

Reply:
Those `_`'d args are used by `storeItems` as it recursively builds a dag from a list of pinned hashes and so are not quite optional but more like private or internal. I've refactored it to make it clear that these are internal pieces of state.